indexd/index/drivers/query/__init__.py

```diff
@@ -0,0 +1,37 @@
+from abc import ABCMeta, abstractmethod
+
+
+class URLsQueryDriver(object):
+    """Relatively abstract class for URLs querying, useful when support for other drivers is added"""
+
+    __metaclass__ = ABCMeta
+
+    @abstractmethod
+    def query_urls(self, exclude=None, include=None, only_versioned=True, offset=0, limit=1000):
+        """ The exclude and include patterns are used to match per record. That is a record wth 3 urls will
+            be returned/excluded if any one of the URLs match the include/exclude patterns
+        Args:
+            exclude (str): url pattern to exclude
+            include (str): url pattern to include
+            only_versioned (bool): query only versioned records or not
+            offset (int):
+            limit (int):
+        Returns:
+            list: result list
+        """
+        raise NotImplementedError("Not Implemented, use a concrete impl")
```

Comment:
There is nothing stopping one from extending a base class and not implementing the required methods. Your IDE will give you warnings if you use one, but that's it. Returning an error makes things more obvious at runtime, just doing a pass might just push the problem to some other section of the code

Reply:
I mean, this actually will not run if a subclass doesn't implement every `abstractmethod`, thanks to `abc`. So the `raise` is unreachable

Reply:
```python
In [1]: from abc import ABCMeta, abstractmethod

In [2]: class Foo(object):
   ...:     __metaclass__ = ABCMeta
   ...:     @abstractmethod
   ...:     def test(self):
   ...:         pass
   ...:     

In [3]: class Bar(Foo):
   ...:     def __init__(self, x):
   ...:         self.x = x
   ...:         

In [4]: b = Bar()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-4-d6290cf35476> in <module>()
----> 1 b = Bar()

TypeError: Can't instantiate abstract class Bar with abstract methods test
```

Reply:
I agree, this is just the behavior of `abc`, like this snippet I posted above if you `raise` instead of `pass` it doesn't matter, it will never get there if the class inherits from an `ABC`.

I will go ahead and walk away from this bike shed now...